File: 08.25/파이참 백업/temp.py
direction = [(0, -1), (-1, 0)]

def dfs(i, j, depth, rr):
    if rr > depth:
        return
    if arr[i][j] == 1:
        visited[i][j] = 1

    for dx, dy in direction:
        for k in range(1, rr + 1):
            ni, nj = i + dx * k, j + dy * k
            if 0 <= ni < N + 1 and 0 <= nj < M and not visited[ni][nj]:
                if arr[ni][nj] == 1:
                    cnt += 1
                    visited[ni][nj] = 1



N, M, D = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
castle = [0] * M
arr.append(castle)

for j in range(M):
    visited = [[0] * M for _ in range(N)]
    bfs(N + 1, j, D, 1)

File: 09.07/algorithm/temp.cpp
#include <iostream>
#include <queue>
#include <tuple>
#include <climits>

using namespace std;

tuple<int, int> direction[] = { {1, 0}, {-1, 0}, {0, 1}, {0, -1} };

int arr[1000][1000];
int visited[1000][1000][2];

int bfs(int N, int M, int si, int sj, int ei, int ej) {
	queue<tuple<int, int, int, int>> q;

	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < M; ++j) {
			visited[i][j][0] = visited[i][j][1] = INT_MAX;
		}
	}

	visited[si][sj][0] = 0;
	q.push({ si, sj, 0, 0 });

	while (!q.empty()) {
		auto[x, y, cnt, wall] = q.front();
		q.pop();

		if (x == ei && y == ej) {
			return cnt;
		}

		for (auto[dx, dy] : direction) {
			int nx = x + dx;
			int ny = y + dy;

			if (0 <= nx && nx < N && 0 <= ny && ny < M) {
				int next_wall = wall + arr[nx][ny];

				if (next_wall <= 1 && visited[nx][ny][next_wall] > cnt + 1) {
					visited[nx][ny][next_wall] = cnt + 1;
					q.push({ nx, ny, cnt + 1, next_wall });
				}
			}
		}
	}

	return -1;
}

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);
	int N, M;

	cin >> N >> M;

	int Hx, Hy;
	cin >> Hx >> Hy;

	int Ex, Ey;
	cin >> Ex >> Ey;

	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < M; ++j) {
			cin >> arr[i][j];
		}
	}

	int result = bfs(N, M, Hx - 1, Hy - 1, Ex - 1, Ey - 1);
	cout << result << endl;

	return 0;
}
